src/shared.ts

```diff
@@ -31,3 +31,130 @@ export function implementsCmp<T>(x: T | Cmp<T>): x is Cmp<T> {
         x.cmp.length === 1
     );
 }
+
+export function defaultEqualityFn<K>(a: K, b: K): boolean {
+    if (implementsEq(a)) {
+        return a.eq(b);
+    }
+    return a === b;
+}
+
+export function defaultComparatorFn<K>(a: K, b: K): Ordering {
+    if (implementsCmp(a)) {
+        return a.cmp(b);
+    }
+    if (a < b) return Ordering.Less;
+    if (a > b) return Ordering.Greater;
+    return Ordering.Equal;
+}
+
+export interface ReadonlySetLike<T> {
+    [Symbol.iterator](): Iterator<T>;
+    size: number;
+    entries(): IterableIterator<[T, T]>;
+    forEach(fn: (value: T) => void): void;
+    has(value: T): boolean;
+    keys(): IterableIterator<T>;
+    values(): IterableIterator<T>;
+}
+
+export interface SetLike<T> extends ReadonlySetLike<T> {
+    add(value: T): this;
+    delete(value: T): boolean;
+    clear(): void;
+}
+
+export interface SetUtils<T> {
+    difference<U>(other: ReadonlySetLike<U>): Set<T>;
+    intersection(other: ReadonlySetLike<T>): Set<T>;
+    isDisjointFrom(other: ReadonlySetLike<T>): boolean;
+    isSubsetOf(other: ReadonlySetLike<T>): boolean;
+    isSupersetOf(other: ReadonlySetLike<T>): boolean;
+    symmetricDifference<U>(other: ReadonlySetLike<U>): Set<T | U>;
+    union<U>(other: ReadonlySetLike<U>): Set<T | U>;
+}
+
+export class Collection<T> extends Set<T> {
```

Comment:
wrong termanology. a `Collection` is an abstract class that a Set should EXTEND.

Maps, Sets, Lists, etc.. are all collections

Reply:
ya i was debating making this abstract, will do that